### PR TITLE
[FrameworkBundle] Remove double cache generation by detecting fresh kernels on cache:clear

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

The `cache:clear` command currently runs non-optional cache warmers twice or more.
It also clears + warms up just created caches (when var/cache/dev doens't exist yet, eg on 1st `composer install`.)

This PR fixes both behaviors. To run only what's needed once.
Best reviewed [ignoring whitespaces](https://github.com/symfony/symfony/pull/26670/files?w=1).

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php [hunk 1]
```diff
@@ -104,34 +104,50 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+                $io->comment('Cache is fresh.');
+            }
+            if (!$input->getOption('no-warmup') && !$input->getOption('no-optional-warmers')) {
+                if ($output->isVerbose()) {
+                    $io->comment('Warming up optional cache...');
+                }
+                $warmer = $kernel->getContainer()->get('cache_warmer');
+                // non optional warmers already ran during container compilation
+                $warmer->enableOnlyOptionalWarmers();
+                $warmer->warmUp($warmupDir);
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php]: shouldn't we create the dir here ?

**Review by stof — CHANGES_REQUESTED** [review]
This looks sensible to me

File: src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php (hunk 1 above)
Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php]: or shouldn't we make it run optional warmers with the actual cache dir directly (which is not this variable), as we won't swap folders ?

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Command/CacheClearCommand.php]: good catch, fixed

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by Tobion** [comment]
I think this is a HUGE DX improvement for many people and many workflows. 

As you said yourself on the first composer install this happens. But it also happens if you have local changes and make a composer update. This runs the cache clear twice basically making it pretty slow esp. if you have time-consuming cache warmers like doc generation.
Or some people just run cache:clear out of habit which often results in doubling the execution time.

IMO we should make this change available at least to SF 3.4 as well.

**Pull request merged**